planner/core/optimizer.go

```diff
@@ -114,12 +119,30 @@ func CheckTableLock(ctx sessionctx.Context, is infoschema.InfoSchema, vs []visit
 	return nil
 }
 
+// LogicalOptimizeTraceBlock records the information what logical rule applies and what's the key impact of each rule.
+type LogicalOptimizeTraceBlock struct {
+	LogicalRules []interface{} `json:"logical_rules"`
+	FinalPlan    string        `json:"final_plan"`
+}
+
 // DoOptimize optimizes a logical plan to a physical plan.
 func DoOptimize(ctx context.Context, flag uint64, logic LogicalPlan) (PhysicalPlan, float64, error) {
+	optTracer := logic.SCtx().GetSessionVars().StmtCtx.OptTracer
+	if optTracer != nil {
+		curBlock := optTracer.TailBlock().(*RootTraceBlock)
+		loBlock := &LogicalOptimizeTraceBlock{}
+		curBlock.ProcessBlock = append(curBlock.ProcessBlock, loBlock)
+		optTracer.AppendBlock(loBlock)
+	}
 	logic, err := logicalOptimize(ctx, flag, logic)
 	if err != nil {
 		return nil, 0, err
 	}
+	if optTracer != nil {
+		curBlock := optTracer.TailBlock().(*LogicalOptimizeTraceBlock)
+		curBlock.FinalPlan = ToString(logic)
```

Comment:
Shall we enhance the `ToString` function this time? This function only shows a few thing about the plan, but we'd better print more details in the tracer.

Reply:
Yes, we may need to do it.
The explain result is not too easy to display in json object since we need to find way to align the explain' text.